File: src/onscala/hackerrank/stacks/BalancedBrackets.scala
package onscala.hackerrank.stacks

import org.scalatest.Assertions
import org.testng.Assert._
import org.testng.annotations.{DataProvider, Test}


/**
  * Site: HackerRank
  * Problem: Balanced Brackets
  *
  * User: Luis Villaca
  * Date: 03/06/2020
  */
class BalancedBrackets extends Assertions {

  val yes = "YES"
  val no = "NO"

  def isBalanced(s: String): String = {

    // 1 - define regular expressions
    val pattern1 = """(.*)(\(\))(.*)""".r // .*().*
    val pattern2 = """(.*)(\[\])(.*)""".r // .*[].*
    val pattern3 = """(.*)(\{\})(.*)""".r // .*{}.*

    // 2 - recursively remove matching pattern from String
      // if balanced, it will be empty in the end of the process
    if (s.isEmpty) yes
    else s match {
      case pattern1 (before,mid,after)  => isBalanced(s"$before$after")
      case pattern2 (before,mid,after)  => isBalanced(s"$before$after")
      case pattern3 (before,mid,after)  => isBalanced(s"$before$after")
      case _ => no
    }
  }

  //Evaluating
  @DataProvider(name= "tests")
  def testData() = {
    Array(
      Array[Any]("[{()}]{}", yes),
      Array[Any]("{[{()}]}", yes),
      Array[Any]("(){}[[]", no),
      Array[Any]("())({}", no)
    )
  }

  @Test(dataProvider = "tests")
  def verify(strInput: String, expectedOutput: String): Unit =
    assertTrue(isBalanced(strInput)  eq expectedOutput)

}

File: src/onjava/hackerrank/stacks/BalancedBrackets.java
package onjava.hackerrank.stacks;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Site: HackerRank
 * Problem: Balanced Brackets
 *
 * User: Luis Villaca
 * Date: 26/04/2020
 */
public class BalancedBrackets {

    public static final String YES = "YES";
    public static final String NO = "NO";

    static String isBalanced(String s) {
        //Brackets properly nested

        Pattern pRound = Pattern.compile("\\(\\)");
        Pattern pCurly = Pattern.compile("\\{\\}");
        Pattern pSquare = Pattern.compile("\\[\\]");

        while (!s.equals("")) {
            String prev = s;
            Matcher matcher = pRound.matcher(s);
            s = matcher.replaceAll("");
            matcher = pCurly.matcher(s);
            s = matcher.replaceAll("");
            matcher = pSquare.matcher(s);
            s = matcher.replaceAll("");
            if (prev.equals(s)) return NO;
        }
        return YES;
    }


    @DataProvider(name = "tests")
    public Object [][] testData() {
        return new Object [][] {
                new Object [] {"[{()}]{}", YES},
                new Object [] {"{[{()}]}", YES},
                new Object [] {"(){}[[]", NO},
                new Object [] {"())({}", NO},
                new Object [] {"{(([])[])[]}[]", YES},
                new Object [] {"{(([])[])[]]}", NO}
        };
    }

    @Test(dataProvider = "tests")
    public void verify(String sInput, String sExpected) {
        String actualOutput = isBalanced(sInput);
        Assert.assertEquals(actualOutput, sExpected);
    }
}
